### DISCUSSION TITLE
Info for pull request #118

### DISCUSSION THREAD

**Discussion by BZanten** [0x6d69636b/windows_hardening#120] [Show and tell]
The idea of this pull request is to turn the 'script' into a real module that can be installed like others.
Also (first) updated some logic in the 'help' output.

1. Updated “help”  in the code

Current code:
```
    .EXAMPLE
        Description: HardeningKitty performs an audit, saves the results and creates a log file:
        Invoke-HardeningKitty -Mode Audit -Log -Report

        Description: HardeningKitty performs an audit with a specific list and does not show machine information:
        Invoke-HardeningKitty -FileFindingList .\lists\finding_list_0x6d69636b_user.csv -SkipMachineInformation

        Description: HardeningKitty ready only the setting with the default list, and saves the results in a specific file:
        Invoke-HardeningKitty -Mode Config -Report -Report C:\tmp\my_hardeningkitty_report.csv
```

Result:

```
PS C:\HardeningKitty\windows_hardening.test3> help Invoke-HardeningKitty -Examples

NAME
    Invoke-HardeningKitty

SYNOPSIS
    Invoke-HardeningKitty - Checks and hardens your Windows configuration


     =^._.^=
    _(      )/  HardeningKitty


    Author:  Michael Schneider
    License: MIT
    Required Dependencies: None
    Optional Dependencies: None


    -------------------------- EXAMPLE 1 --------------------------

    PS C:\>Description: HardeningKitty performs an audit, saves the results and creates a log file:

    Invoke-HardeningKitty -Mode Audit -Log -Report

    Description: HardeningKitty performs an audit with a specific list and does not show machine information:
    Invoke-HardeningKitty -FileFindingList .\lists\finding_list_0x6d69636b_user.csv -SkipMachineInformation

    Description: HardeningKitty ready only the setting with the default list, and saves the results in a specific file:
    Invoke-HardeningKitty -Mode Config -Report -Report C:\tmp\my_hardeningkitty_report.csv
```
Note:
•	even though you provided 3 examples, only 1 example ' ------ EXAMPLE 1 -------'  is outputted
•	there is a PS C:\>    added before your 'Description'  that is not logical
•	The FIRST line after the '.EXAMPLE' should be the line with code... (PowerShell will pre-prend this with PS C:\> )
•	(All) additional lines after that are treated as, and should be, the description of that example

Updated code for the .Example part:

```
    .EXAMPLE 
        Invoke-HardeningKitty -Mode Audit -Log -Report

        HardeningKitty performs an audit, saves the results and creates a log file

    .EXAMPLE
        Invoke-HardeningKitty -FileFindingList .\lists\finding_list_0x6d69636b_user.csv -SkipMachineInformation

        HardeningKitty performs an audit with a specific list and does not show machine information

    .EXAMPLE
        Invoke-HardeningKitty -Mode Config -Report -Report C:\tmp\my_hardeningkitty_report.csv

        HardeningKitty ready only the setting with the default list, and saves the results in a specific file
```
So 3x .Example, for each example start with the code, then the explanation

Result

```
PS C:\WINDOWS\system32> Import-module HardeningKitty 
PS C:\WINDOWS\system32> get-help Invoke-HardeningKitty -Examples

NAME
    Invoke-HardeningKitty

SYNOPSIS
    Invoke-HardeningKitty - Checks and hardens your Windows configuration


     =^._.^=
    _(      )/  HardeningKitty


    Author:  Michael Schneider
    License: MIT
    Required Dependencies: None
    Optional Dependencies: None


    -------------------------- EXAMPLE 1 --------------------------

    PS C:\>Invoke-HardeningKitty -Mode Audit -Log -Report

    HardeningKitty performs an audit, saves the results and creates a log file


    -------------------------- EXAMPLE 2 --------------------------

    PS C:\>Invoke-HardeningKitty -FileFindingList .\lists\finding_list_0x6d69636b_user.csv -SkipMachineInformation

    HardeningKitty performs an audit with a specific list and does not show machine information


    -------------------------- EXAMPLE 3 --------------------------

    PS C:\>Invoke-HardeningKitty -Mode Config -Report -Report C:\tmp\my_hardeningkitty_report.csv

    HardeningKitty ready only the setting with the default list, and saves the results in a specific file
```
Note that now, there are 3 separate examples, as expected, and each example commandline is pre-fixed with PS C:\>

2. Turned the .PS1 script into a module
   a. Rename .ps1 to .psm1
  I took the liberty to rename the file itself from Invoke-HardeningKitty.ps1  to HardeningKitty   because I'm assuming here the PS Module should be called HardeningKitty and the function to invoke is the Invoke-*  one.
   b. add Export-ModuleMember to the end of the script
Export-ModuleMember -Function Invoke-HardeningKitty
   c. Add a .psd1 script, to describe the metadata of this script.
   d. Because this rename is quite impactful I also took the liberty to bump the version from 0.8 to 0.9  you may even start to consider it creating it a 1.0  but let's get all in place first.
Note!
you may want to review and update the metadata, especially regarding your names and required license etc.  (the GUID is a random new guid generated via New-GUID)

3. Tested the Module
   a. by placing it in the folder:
      C:\Program Files\WindowsPowerShell\Modules\HardeningKitty\0.9
   b. run the script:
```
   Import-Module HardeningKitty
   Invoke-HardeningKitty
```
   c. Resulting in error:
```
[!] 8/27/2022 6:29:02 PM - The finding list C:\\lists\finding_list_0x6d69636b_machine.csv was not found.
```
      This is because it is now a Module, it is not normal to 'CD' into the folder where the module is loaded from, so the Invoke- command can't find the input csv file

   d. Update the code, in 2 places
      `$CurrentLocation = Get-Location`
      to
      `$CurrentLocation = $PSScriptRoot`
      because $PSScriptRoot is the root of the module that is running, thus the default vulnerability lists are relative to the installed module.

**Comment by 0x6d69636b** [comment]
Thank you very much for your work. I'll look into it in the next days. I already created the "last" release of the script version, so before we change to a module people can get the current version.

**Comment by BZanten** [comment]
Hi Michael
Yes thanks.  Below is the text of what I’ve done.
If my changes interfere with other changes, then there is no problem to ignore the pull request, and we can easily re-do the updates in this pull request.

I also just updated the version number, no idea what it should become.
PS regarding version number: you had it in your .ps1  script,  - now  the .psm1 .   but  the .psd1  script ALSO  specifies the version number (of the whole module itself) you may want to consider changing the version number in both files in future changes.

I have NOT researched how this module can finally be published to the PowerShell gallery or any other gallery where you’d like to have it.
Builtin is of course the PSGallery:
PS C:\Users\ZantenBAM> Get-PSRepository

Name                      InstallationPolicy   SourceLocation
----                      ------------------   --------------
PSGallery                 Untrusted            https://www.powershellgallery.com/api/v2

Let’s look at that a later time.

Regards, Ben van Zanten



  1.  Update “help”  in the code
Current code:

    .EXAMPLE
        Description: HardeningKitty performs an audit, saves the results and creates a log file:
        Invoke-HardeningKitty -Mode Audit -Log -Report

        Description: HardeningKitty performs an audit with a specific list and does not show machine information:
        Invoke-HardeningKitty -FileFindingList .\lists\finding_list_0x6d69636b_user.csv -SkipMachineInformation

        Description: HardeningKitty ready only the setting with the default list, and saves the results in a specific file:
        Invoke-HardeningKitty -Mode Config -Report -Report C:\tmp\my_hardeningkitty_report.csv

Result:

PS C:\HardeningKitty\windows_hardening.test3> help Invoke-HardeningKitty -Examples

NAME
    Invoke-HardeningKitty

SYNOPSIS
    Invoke-HardeningKitty - Checks and hardens your Windows configuration


     =^._.^=
    _(      )/  HardeningKitty


    Author:  Michael Schneider
    License: MIT
    Required Dependencies: None
    Optional Dependencies: None


    -------------------------- EXAMPLE 1 --------------------------

    PS C:\>Description: HardeningKitty performs an audit, saves the results and creates a log file:

    Invoke-HardeningKitty -Mode Audit -Log -Report

    Description: HardeningKitty performs an audit with a specific list and does not show machine information:
    Invoke-HardeningKitty -FileFindingList .\lists\finding_list_0x6d69636b_user.csv -SkipMachineInformation

    Description: HardeningKitty ready only the setting with the default list, and saves the results in a specific file:
    Invoke-HardeningKitty -Mode Config -Report -Report C:\tmp\my_hardeningkitty_report.csv


Note:

  *   even though you provided 3 examples, only 1 example ' ------ EXAMPLE 1 -------'  is outputted
  *   there is a PS C:\>    added before your 'Description'  that is not logical
  *   The FIRST line after the '.EXAMPLE' should be the line with code... (PowerShell will pre-prend this with PS C:\> )
  *   (All) additional lines after that are treated as, and should be, the description of that example
Updated code for the .Example part:

    .EXAMPLE
        Invoke-HardeningKitty -Mode Audit -Log -Report

        HardeningKitty performs an audit, saves the results and creates a log file

    .EXAMPLE
        Invoke-HardeningKitty -FileFindingList .\lists\finding_list_0x6d69636b_user.csv -SkipMachineInformation

        HardeningKitty performs an audit with a specific list and does not show machine information

    .EXAMPLE
        Invoke-HardeningKitty -Mode Config -Report -Report C:\tmp\my_hardeningkitty_report.csv

        HardeningKitty ready only the setting with the default list, and saves the results in a specific file


Result

PS C:\WINDOWS\system32> Import-module HardeningKitty
PS C:\WINDOWS\system32> get-help Invoke-HardeningKitty -Examples

NAME
    Invoke-HardeningKitty

SYNOPSIS
    Invoke-HardeningKitty - Checks and hardens your Windows configuration


     =^._.^=
    _(      )/  HardeningKitty


    Author:  Michael Schneider
    License: MIT
    Required Dependencies: None
    Optional Dependencies: None


    -------------------------- EXAMPLE 1 --------------------------

    PS C:\>Invoke-HardeningKitty -Mode Audit -Log -Report

    HardeningKitty performs an audit, saves the results and creates a log file


    -------------------------- EXAMPLE 2 --------------------------

    PS C:\>Invoke-HardeningKitty -FileFindingList .\lists\finding_list_0x6d69636b_user.csv -SkipMachineInformation

    HardeningKitty performs an audit with a specific list and does not show machine information


    -------------------------- EXAMPLE 3 --------------------------

    PS C:\>Invoke-HardeningKitty -Mode Config -Report -Report C:\tmp\my_hardeningkitty_report.csv

    HardeningKitty ready only the setting with the default list, and saves the results in a specific file


Note that now, there are 3 separate examples, as expected, and each example commandline is pre-fixed with PS C:\>

  1.   Turn the .PS1 script into a module
     *   Rename .ps1 to .psm1
I took the liberty to rename the file itself from Invoke-HardeningKitty.ps1  to HardeningKitty   because I'm assuming here the PS Module should be called HardeningKitty and the function to invoke is the Invoke-*  one.
     *   add Export-ModuleMember to the end of the script
Export-ModuleMember -Function Invoke-HardeningKitty

     *   Add a .psd1 script, to describe the metadata of this script.
     *   Because this rename is quite impactful I also took the liberty to bump the version from 0.8 to 0.9  you may even start to consider it creating it a 1.0  but let's get all in place first.
Note!
you may want to review and update the metadata, especially regarding your names and required license etc.  (the GUID is a random new guid generated via New-GUID)


  1.  Test the Module
     *   by placing it in the folder:
C:\Program Files\WindowsPowerShell\Modules\HardeningKitty\0.9
     *   run the script:
PS C:\> Import-Module HardeningKitty
PS C:\> Invoke-HardeningKitty
     *   Resulting in error:

[!] 8/27/2022 6:29:02 PM - The finding list C:\\lists\finding_list_0x6d69636b_machine.csv was not found.

This is because it is now a Module, it is not normal to 'CD' into the folder where the module is loaded from, so the Invoke- command can't find the input csv file
     *   Update the code, in 2 places
$CurrentLocation = Get-Location
to
$CurrentLocation = $PSScriptRoot
because $PSScriptRoot is the root of the module that is running, thus the default vulnerability lists are relative to the installed module.
